R/compare.R

```diff
@@ -56,9 +56,43 @@
 #' @param ignore_encoding Ignore string encoding? `TRUE` by default, because
 #'   this is R's default behaviour. Use `FALSE` when specifically concerned
 #'   with the encoding, not just the value of the string.
+#' @param ignore_name_order Ignore changes to the order of named components in
+#'   lists or vectors.
+#' @param ignore_NULLs In lists and pairlists, ignore elements that are set to
```

Comment:
Is the motivation for this to treat `list(a = 1, b = NULL)` and `list(a = 1)` as equivalent?  If so, I wonder if we could possibly combine `ignore_name_order` and `ignore_NULLs` into some more generic option like `list_as_dictionary = TRUE`. The idea of "list_as_dictionary" is to imply that we're comparing lists based on the results of named subsetting. I'm open to better names if you think combining these two options might make sense.

Reply:
Yes, the `ignore_NULLs` is for that sort of difference.

I think merging these is a matter of taste:  how fine-grained do you want the control to be?  Having `ignore_name_order` forced equal to `ignore_NULLs` would suit my uses.  

The name `list_as_dictionary` would be okay for that merged option for sophisticated users, but might confuse users who know R but not other languages that use the name "dictionary".   However, I can't think of a better one.

Reply:
Another option is to copy `expect_mapequal()` and do `list_as_map`; I don't think that's much better in terms of recognising the term ("map"), but at least it connects to an existing function.

Reply:
What should happen if there are duplicate names? If the principle is to match the behaviour of `[[`, then I guess only the first instance of the name should be considered?

Reply:
Or maybe that's not worth handling specially and will just show a difference.

Reply:
I'd say duplicate names are uncommon enough that anyone who has them probably wants them, so I wouldn't remove dups.  That is, if the names are `c("b", "a", "b")` and I ask to ignore name order, I'd want it to compare equal to `c("a", "b", "b")` but not equal to `c("a", "b")`.